rust/mdb_shard/src/intershard_reference_structs.rs

```diff
@@ -0,0 +1,279 @@
+                let total_dedup_hit_count: u32 = count.try_into().unwrap_or(u32::MAX);
+
+                IntershardReferenceSequenceEntry {
+                    shard_hash,
+                    total_dedup_chunks: total_dedup_hit_count,
+                }
+            }));
+
+        entries.sort_unstable_by_key(|e| u64::MAX - e.total_dedup_chunks as u64);
+
+        if entries.len() > INTERSHARD_REFERENCE_SIZE_CAP {
+            entries.resize(INTERSHARD_REFERENCE_SIZE_CAP, Default::default());
+        }
+
+        Self {
+            metadata: IntershardReferenceSequenceHeader::new(entries.len()),
+            entries,
+        }
+    }
+
+    pub fn num_bytes(&self) -> u64 {
+        (size_of::<IntershardReferenceSequenceHeader>()
+            + self.entries.len() * size_of::<IntershardReferenceSequenceEntry>()) as u64
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.entries.is_empty()
+    }
+
+    pub fn merge(self, other: IntershardReferenceSequence) -> Self {
+        let mut s = self;
+
+        let entries = take(&mut s.entries);
+        let mut local_hm: HashMap<MerkleHash, IntershardReferenceSequenceEntry> = entries
+            .into_iter()
+            .map(|irse| (irse.shard_hash, irse))
+            .collect();
+
+        for irse in other.entries.into_iter() {
+            let entry = local_hm
+                .entry(irse.shard_hash)
+                .or_insert_with(|| IntershardReferenceSequenceEntry::new(irse.shard_hash, 0));
+            entry.total_dedup_chunks = entry
+                .total_dedup_chunks
+                .saturating_add(irse.total_dedup_chunks);
+        }
+
+        // Collect the entries at the end.
+        s.entries = local_hm.into_values().collect();
+
+        // Sort them in reverse order by number of hits
+        s.entries
+            .sort_unstable_by_key(|e| u64::MAX - (e.total_dedup_chunks as u64));
+
+        if s.entries.len() > INTERSHARD_REFERENCE_SIZE_CAP {
+            s.entries
+                .resize(INTERSHARD_REFERENCE_SIZE_CAP, Default::default());
+        }
+
+        s
+    }
+
+    pub fn serialize<W: Write>(
+        &self,
+        writer: &mut W,
+    ) -> std::result::Result<usize, std::io::Error> {
+        let mut n_bytes = 0;
+
+        n_bytes += self.metadata.serialize(writer)?;
+
+        for isre in self.entries.iter() {
+            n_bytes += isre.serialize(writer)?;
+        }
+
+        Ok(n_bytes)
+    }
+
+    pub fn deserialize<R: Read>(reader: &mut R) -> std::result::Result<Self, std::io::Error> {
+        let metadata = IntershardReferenceSequenceHeader::deserialize(reader)?;
+
+        let mut entries = Vec::with_capacity(metadata.num_entries as usize);
+        for _ in 0..metadata.num_entries {
+            entries.push(IntershardReferenceSequenceEntry::deserialize(reader)?);
+        }
+
+        Ok(Self { metadata, entries })
+    }
+}
+
+pub fn write_out_with_new_intershard_reference_section<R: Read + Seek>(
```

Comment:
And is this function used?

Reply:
Again, not in this PR; it will be in the dependent PR after this.